python/ray/autoscaler/_private/resource_demand_scheduler.py

```diff
@@ -145,6 +163,47 @@ def add_node(node_type, available_resources=None):
 
         return node_resources, node_type_counts
 
+    def reserve_and_allocate_spread(self,
+                                    strict_spreads: List[List[ResourceDict]],
+                                    node_resources: List[ResourceDict],
+                                    node_type_counts: Dict[NodeType, int]):
+        """For each strict spread, attempt to reserve as much space as possible on the
+           node, then allocate new nodes for the unfulfilled portion.
+
+        Args:
+            strict_spreads (List[List[ResourceDict]]): A list of placement
+                groups which must be spread out.
+            node_resources (List[ResourceDict]): Available node resources in
+                the cluster.
+            node_type_counts (Dict[NodeType, int]): The amount of each type of node
+                pending or in the cluster.
+
+        Returns:
+            Dict[NodeType, int]: Nodes to add.
+            List[ResourceDict]: The updated node_resources after the method.
+            Dict[NodeType, int]: The updated node_type_counts.
+        """
+        to_add = collections.defaultdict(int)
+        for bundles in strict_spreads:
+            unfulfilled, node_resources = get_bin_pack_residual(
+                node_resources, bundles, strict_spread=True)
+            max_to_add = self.max_workers - sum(node_type_counts.values())
+            to_launch = get_nodes_for(
+                self.node_types,
+                node_type_counts,
+                max_to_add,
+                unfulfilled,
+                strict_spread=True)
+            _inplace_add(node_type_counts, to_launch)
+            _inplace_add(to_add, to_launch)
+            new_node_resources = _node_type_counts_to_node_resources(
+                self.node_types, to_launch)
+            unfulfilled, including_reserved = get_bin_pack_residual(
+                new_node_resources, unfulfilled, strict_spread=True)
+            assert not unfulfilled
+            node_resources += including_reserved
+        return to_add, node_resources, node_type_counts
```

Comment:
This function is pretty hard to understand. Consider adding more comments with a running example to clarify it to the reader.

Reply:
I've added some comments, but not really sure about a running example. There is an example in the form of unit testing though (which is nice because it's guaranteed to be up to date).